crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
-                });
-                line = &line[code.len()..];
 
-                // Codes can be comma- or whitespace-delimited.
-                if let Some(rest) = Self::lex_delimiter(line).map(Self::lex_whitespace) {
-                    line = rest;
-                } else {
-                    break;
-                }
+            self.cursor.skip_bytes(tool_len);
+
+            // `#\s*(?:ruff|flake8)\s*:\s*(?i)noqa`
+            //                     ^^^
+            self.cursor.eat_while(char::is_whitespace);
+
+            // `#\s*(?:ruff|flake8)\s*:\s*(?i)noqa`
+            //                        ^
+            if !self.cursor.eat_char(':') {
+                continue;
             }
 
-            // If we didn't identify any codes, warn.
-            if codes.is_empty() {
-                return Err(ParseError::MissingCodes);
+            // `#\s*(?:ruff|flake8)\s*:\s*(?i)noqa`
+            //                         ^^^
+            self.cursor.eat_while(char::is_whitespace);
+
+            // `#\s*(?:ruff|flake8)\s*:\s*(?i)noqa`
+            //                             ^^^^^^^
+            if !is_noqa_uncased(self.cursor.as_str()) {
+                continue;
             }
+            self.cursor.skip_bytes("noqa".len());
 
-            let codes_end = init_line_len - line.text_len();
-            let range = TextRange::new(comment_start, codes_end);
-            Self::Codes(Codes {
-                range: range.add(offset),
-                codes,
-            })
-        }))
+            return Ok(Some(self.lex_directive()?));
+        }
+        Ok(None)
     }
 
-    /// Lex optional leading whitespace.
-    #[inline]
-    fn lex_whitespace(line: &str) -> &str {
-        line.trim_start()
+    /// Collect codes in `noqa` comment.
+    fn lex_directive(mut self) -> Result<NoqaLexerOutput<'a>, LexicalError> {
+        match self.cursor.bump() {
+            None => {
+                let range = TextRange::at(self.offset, self.current());
+                return Ok(NoqaLexerOutput {
+                    warnings: self.warnings,
+                    directive: Directive::All(All { range }),
+                });
+            }
+            // Ex) # noqa# A comment
+            //            ^
+            Some('#') => {
+                let range = TextRange::at(self.offset, self.current() - '#'.text_len());
+                return Ok(NoqaLexerOutput {
+                    warnings: self.warnings,
+                    directive: Directive::All(All { range }),
+                });
+            }
+            // Ex) # noqa A comment
+            Some(c) if c.is_whitespace() => {
+                let range = TextRange::at(self.offset, self.current() - c.text_len());
+                return Ok(NoqaLexerOutput {
+                    warnings: self.warnings,
+                    directive: Directive::All(All { range }),
+                });
+            }
+            // Ex) # noqa: F401,F842
```

Comment:
I'm wondering if we should be more lenient. E.g. `noqa : F401` will pass without any error but a user clearly meant to write `noqa: F401`. Should we just skip any whitespace after `noqa` and then decide on the next character?